src/lpython/parser/tokenizer.re

```diff
@@ -397,6 +397,10 @@ int Tokenizer::lex(Allocator &al, YYSTYPE &yylval, Location &loc, diag::Diagnost
             "yield"    { KW(YIELD) }
             "yield" whitespace "from" whitespace { KW(YIELD_FROM) }
 
+            // Soft Keywords
+            "match" / [^\n\x00]+ ":" newline { KW(MATCH) }
+            "case" / [^\n\x00]+ ":" newline { KW(CASE) }
```

Comment:
I think to fix the test cases that I posted above, you need to check here that the last token was TK_NEWLINE, if so, it's a KW_MATCH, otherwise it's a TK_NAME.